cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,452 @@
+
+    auto left_end_zip_iterator = thrust::make_zip_iterator(thrust::make_tuple(left_side + left_size, left_indices + left_size));
+    auto right_end_zip_iterator = thrust::make_zip_iterator(thrust::make_tuple(right_side + right_size, right_indices + right_size));
+
+    rmm::device_vector<index_type> merged_indices(total_size);
+    
+    auto lhs_tdv = cudf::table_device_view::create(left_table, stream);
+    auto rhs_tdv = cudf::table_device_view::create(right_table, stream);
+
+    rmm::device_vector<cudf::order> d_column_order(asc_desc); 
+    
+    auto exec_pol = rmm::exec_policy(stream);
+    if (nullable){
+      rmm::device_vector<cudf::null_order> d_null_precedence(null_precedence);
+      
+      auto ineq_op =
+        cudf::experimental::row_lexicographic_comparator<true>(*lhs_tdv,
+                                                               *rhs_tdv,
+                                                               d_column_order.data().get(),
+                                                               d_null_precedence.data().get());
+      
+        thrust::merge(exec_pol->on(stream),
+                    left_begin_zip_iterator,
+                    left_end_zip_iterator,
+                    right_begin_zip_iterator,
+                    right_end_zip_iterator,
+                    merged_indices.begin(),
+                    [=] __device__ (thrust::tuple<side, cudf::size_type> const & right_tuple,
+                                    thrust::tuple<side, cudf::size_type> const & left_tuple) {
+                        return ineq_op(thrust::get<1>(right_tuple), thrust::get<1>(left_tuple));
+                    });			        
+    } else {
+      auto ineq_op =
+        cudf::experimental::row_lexicographic_comparator<false>(*lhs_tdv,
+                                                                *rhs_tdv,
+                                                                d_column_order.data().get()); 
+        thrust::merge(exec_pol->on(stream),
+                    left_begin_zip_iterator,
+                    left_end_zip_iterator,
+                    right_begin_zip_iterator,
+                    right_end_zip_iterator,
+                    merged_indices.begin(),
+                    [=] __device__ (thrust::tuple<side, cudf::size_type> const & right_tuple,
+                                    thrust::tuple<side, cudf::size_type> const & left_tuple) {
+                        return ineq_op(thrust::get<1>(right_tuple), thrust::get<1>(left_tuple));
+                    });					        
+    }
+
+    CHECK_STREAM(stream);
+
+    return merged_indices;
+}
+
+} // namespace
+
+namespace cudf {
+namespace experimental { 
+namespace detail {
+
+//work-in-progress:
+//
+//generate merged column
+//given row order of merged tables
+//(ordered according to indices of key_cols)
+//and the 2 columns to merge
+//
+template<typename VectorI>
```

Comment:
The idea was that `ColumnMerger` might get (eventually) exposed for other uses than just merge. In which case it might be useful to accept different types of containers. 

Or, could be specialized for containers of things other than `index_type`; e.g., if the "extended index" logic might be used instead (ref. to m-b-k discussion w/ Trevor). 

But I can certainly take it out.

Reply:
Ok, removed it.